### PR TITLE
Save ConvertTransform as ONNX Operator and Control the Use of Experimental Features with a Flag

### PR DESCRIPTION
**Pull request by wschin**
1. Introduce a new argument to SaveOnnx, which is OnnxVersion.
   Two values are currently allowed, "Latest" and "Experimental".
   Note that "Latest" means that the produced ONNX model meets
   the latest ONNX release while "Experimental" may produce
   things not officially supported in ONNX.
2. For (1), the interface of saving ONNX is slightly changed. Now,
   CanSaveOnnx requires an OnnxContext as its input argument,
   because if a model can be saved to ONNX depends on the targeted
   ONNX version now.
3. Add exporter for ConvertTransform. It doesn't use standard
   ONNX operator.

Related issue: #945

### CODE REVIEW

**Comment by wschin** [comment]
@markusweimer , sure. Thanks for reminding.

File: src/Microsoft.ML.Data/Model/Onnx/OnnxContext.cs [hunk 1]
```diff
@@ -7,6 +7,8 @@
 
 namespace Microsoft.ML.Runtime.Model.Onnx
 {
+    public enum OnnxVersion { Latest, Experimental }
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Data/Model/Onnx/OnnxContext.cs]: >public enum OnnxVersion { Latest, Experimental  [](start = 4, length = 47)

This doesn't feel right.
First of all, I would expect not enum, but some kind of string.
Second why Latest version is first in list?
Third, you need to assign values.

Comment (wschin) [reply src/Microsoft.ML.Data/Model/Onnx/OnnxContext.cs]: Hello,
(1) I was using string but then agreed with @TomFinley that string is ambigous and the number of supported versions is likely tiny.
(2) Maybe change Latest to Stable could be better? I assume that the first option is the option a user may prefer.
(3) What do you mean by assign value? It's an enum declaration.
Thanks,


---
In reply to: [219627520](https://github.com/dotnet/machinelearning/pull/947#discussion_r219627520) [](ancestors = 219627520)

Comment (Ivanidzo4ka) [reply src/Microsoft.ML.Data/Model/Onnx/OnnxContext.cs]: 3) public enum OnnxVersion: byte { Stable=0, Experimental=1} 

---
In reply to: [219651774](https://github.com/dotnet/machinelearning/pull/947#discussion_r219651774) [](ancestors = 219651774,219627520)

**Comment by Ivanidzo4ka** [comment]
Can you make sure it pass build system?

File: src/Microsoft.ML.Legacy/CSharpApi.cs [hunk 2]
```diff
@@ -3309,6 +3315,11 @@ public sealed partial class OnnxConverter
             /// </summary>
             public Var<Microsoft.ML.Runtime.EntryPoints.ITransformModel> Model { get; set; } = new Var<Microsoft.ML.Runtime.EntryPoints.ITransformModel>();
 
+            /// <summary>
+            /// The targeted ONNX version. It can be either "Stable" or "Experimental"
```
Comment (markusweimer) [inline src/Microsoft.ML.Legacy/CSharpApi.cs]: What does Stable and Experimental refer to? Should this be documented here?

Comment (wschin) [reply src/Microsoft.ML.Legacy/CSharpApi.cs]: Will do. Thanks.

File: src/Microsoft.ML.Onnx/OnnxUtils.cs [hunk 3]
```diff
@@ -308,14 +308,38 @@ public ModelArgs(string name, TensorProto.Types.DataType dataType, List<long> di
...
+                    dataType = TensorProto.Types.DataType.Int16;
+                    break;
+                case DataKind.U2:
+                    dataType = TensorProto.Types.DataType.Uint16;
+                    break;
+                case DataKind.I4:
+                    dataType = TensorProto.Types.DataType.Int32;
+                    break;
                 case DataKind.U4:
                     dataType = TensorProto.Types.DataType.Int64;
```
Comment (justinormont) [inline src/Microsoft.ML.Onnx/OnnxUtils.cs]: Should this be a `Uint32`? Or is there no support for a Uint32, and we're storing in a Int64?

@TomFinley

File: src/Microsoft.ML.Onnx/SaveOnnxCommand.cs [hunk 4]
```diff
@@ -57,6 +57,9 @@ public sealed class Arguments : DataCommand.ArgumentsBase
 
             [Argument(ArgumentType.Required, Visibility = ArgumentAttribute.VisibilityType.EntryPointsOnly, HelpText = "Model that needs to be converted to ONNX format.", SortOrder = 10)]
             public ITransformModel Model;
+
+            [Argument(ArgumentType.AtMostOnce, HelpText = "The targeted ONNX version. It can be either \"Stable\" or \"Experimental\". If \"Experimentab\" is used, models produced can contain components not officially defined in ONNX standard.", SortOrder = 11)]
```
Comment (justinormont) [inline src/Microsoft.ML.Onnx/SaveOnnxCommand.cs]: `Experimentab` should be `Experimental`

Comment (justinormont) [reply src/Microsoft.ML.Onnx/SaveOnnxCommand.cs]: `Experimentab` is in core_manifest.json

File: src/Microsoft.ML.Onnx/OnnxUtils.cs (hunk 3 above)
Comment (wschin) [reply src/Microsoft.ML.Onnx/OnnxUtils.cs]: This is not a part of this PR. That mapping is used across the last two Windows ML releases.

File: src/Microsoft.ML.Onnx/SaveOnnxCommand.cs (hunk 4 above)
Comment (wschin) [reply src/Microsoft.ML.Onnx/SaveOnnxCommand.cs]: Good catch!

File: src/Microsoft.ML.Data/Model/Onnx/OnnxContext.cs (hunk 1 above)
Comment (wschin) [reply src/Microsoft.ML.Data/Model/Onnx/OnnxContext.cs]: Done!

File: src/Microsoft.ML.Onnx/OnnxUtils.cs [hunk 5]
```diff
@@ -311,14 +311,38 @@ public ModelArgs(string name, TensorProto.Types.DataType dataType, List<long> di
...
+                    break;
                 case DataKind.R4:
                     dataType = TensorProto.Types.DataType.Float;
                     break;
+                case DataKind.R8:
+                    dataType = TensorProto.Types.DataType.Double;
+                    break;
                 default:
-                    Contracts.Assert(false, "Unknown type.");
+                    Contracts.Assert(false, "Unsupported type: DataKind " + rawKind.ToString());
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Onnx/OnnxUtils.cs]: I would advise to use Check instead of Assert

File: src/Microsoft.ML.Onnx/SaveOnnxCommand.cs [hunk 6]
```diff
@@ -57,6 +57,9 @@ public sealed class Arguments : DataCommand.ArgumentsBase
 
             [Argument(ArgumentType.Required, Visibility = ArgumentAttribute.VisibilityType.EntryPointsOnly, HelpText = "Model that needs to be converted to ONNX format.", SortOrder = 10)]
             public ITransformModel Model;
+
+            [Argument(ArgumentType.AtMostOnce, HelpText = "The targeted ONNX version. It can be either \"Stable\" or \"Experimental\". If \"Experimental\" is used, models produced can contain components not officially defined in ONNX standard.", SortOrder = 11)]
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Onnx/SaveOnnxCommand.cs]: >models produced can contain components not officially defined in ONNX standard [](start = 164, length = 78)

is used, produced model can contain components that is not officially supported in ONNX standard.

File: test/Microsoft.ML.Tests/OnnxTests.cs [hunk 7]
```diff
@@ -59,7 +59,7 @@ public void InitializerCreationTest()
             using (var env = new ConsoleEnvironment())
             {
                 // Create the actual implementation
-                var ctxImpl = new OnnxContextImpl(env, "model", "ML.NET", "0", 0, "com.test");
+                var ctxImpl = new OnnxContextImpl(env, "model", "ML.NET", "0", 0, "com.test", Runtime.Model.Onnx.OnnxVersion.Stable);
```
Comment (Ivanidzo4ka) [inline test/Microsoft.ML.Tests/OnnxTests.cs]: >Runtime.Model.Onnx [](start = 94, length = 18)

it's weird what you can just type OnnxContextImpl but for 
version you need to provide whole namespace, aren't they belong to same namespace?

File: src/Microsoft.ML.Data/Transforms/ConvertTransform.cs [hunk 8]
```diff
@@ -375,6 +376,30 @@ public override void Save(ModelSaveContext ctx)
...
+        protected override bool SaveAsOnnxCore(OnnxContext ctx, int iinfo, ColInfo info, string srcVariableName, string dstVariableName)
+        {
+            var opType = "CSharp";
+
+            for (int i = 0; i < _exes.Length; i++)
+            {
+                var ex = _exes[i];
+                var node = ctx.CreateNode(opType, srcVariableName, dstVariableName, ctx.GetNodeName(opType));
+                node.AddAttribute("type", LoaderSignature);
+                node.AddAttribute("to", (byte)ex.Kind);
```
[the remark this answers is not in the document]
Comment (wschin) [reply src/Microsoft.ML.Data/Transforms/ConvertTransform.cs]: It's a CSharp operator which is not officially included in ONNX. We can define whatever we need and our runtime will be able to handle it.

---
In reply to: [223531409](https://github.com/dotnet/machinelearning/pull/947#discussion_r223531409) [](ancestors = 223531409)

File: src/Microsoft.ML.Onnx/OnnxUtils.cs (hunk 3 above)
Comment (justinormont) [reply src/Microsoft.ML.Onnx/OnnxUtils.cs]: I opened an issue to discuss/track: https://github.com/dotnet/machinelearning/issues/1198

File: src/Microsoft.ML.Onnx/OnnxUtils.cs (hunk 5 above)
Comment (wschin) [reply src/Microsoft.ML.Onnx/OnnxUtils.cs]: Done.

---
In reply to: [223532122](https://github.com/dotnet/machinelearning/pull/947#discussion_r223532122) [](ancestors = 223532122)

File: test/Microsoft.ML.Tests/OnnxTests.cs (hunk 7 above)
Comment (wschin) [reply test/Microsoft.ML.Tests/OnnxTests.cs]: ONNXVersion appears at two different places, legacy C# API and the main ONNX assembly. Unfortunately, both of them are included in this test.

---
In reply to: [223532859](https://github.com/dotnet/machinelearning/pull/947#discussion_r223532859) [](ancestors = 223532859)

**Comment by justinormont** [comment]
Looks like there's a test which failed: 
```
2018-10-09T17:45:52.5177105Z Failed   Microsoft.ML.Runtime.RunTests.TestEntryPoints.EntryPointCatalog
2018-10-09T17:45:52.5177853Z Error Message:
2018-10-09T17:45:52.5178689Z  Assert.True() Failure
2018-10-09T17:45:52.5178806Z Expected: True
2018-10-09T17:45:52.5179047Z Actual:   False
2018-10-09T17:45:52.5179125Z Stack Trace:
2018-10-09T17:45:52.5179196Z    at Microsoft.ML.Runtime.RunTests.BaseTestBaseline.Done() in /__w/1/s/test/Microsoft.ML.TestFramework/BaseTestBaseline.cs:line 149
2018-10-09T17:45:52.5179510Z    at Microsoft.ML.Runtime.RunTests.TestEntryPoints.EntryPointCatalog() in /__w/1/s/test/Microsoft.ML.Core.Tests/UnitTests/TestEntryPoints.cs:line 300
2018-10-09T17:45:52.5180265Z Standard Output Messages:
2018-10-09T17:45:52.5181202Z  Output matches baseline: '../Common/EntryPoints/core_ep-list.tsv'
2018-10-09T17:45:52.5181605Z  *** Failure: Output and baseline mismatch at line 2492: '../Common/EntryPoints/core_manifest.json'
2018-10-09T17:45:52.5181707Z  Test EntryPointCatalog: completed normally: failed
2018-10-09T17:45:52.5181824Z 
2018-10-09T17:45:52.5182157Z 
2018-10-09T17:45:52.5709539Z Results File: /__w/1/s/bin/AnyCPU.Debug/Microsoft.ML.Core.Tests/_d036fad37c59_2018-10-09_17_44_56.trx
2018-10-09T17:45:52.5712294Z 
2018-10-09T17:45:52.5715346Z Total tests: 105. Passed: 101. Failed: 1. Skipped: 3.
2018-10-09T17:45:52.5717716Z Test Run Failed.
2018-10-09T17:45:52.5734548Z Test execution time: 59.2323 Seconds
```

**Pull request merged**